continue Richard's PR #2

66eli77 opened the pull request:
link to original: #2

rtibbles committed:
Load data based on current mastery models and attempt logs.

66eli77 committed:
kebab-case fun

66eli77 committed:
attempts cannot be undefined

66eli77 committed:
extra '/'

66eli77 committed:
always catch the reject for easy debugging

66eli77 committed:
don't know where should call these 2 methods yet. Temporally call them here

66eli77 committed:
comply to new changes in kolibri master

66eli77 committed:
comply new changes in kolibri master

66eli77 committed:
the replacing path_holder login is moved to kolibri ZipContentView

66eli77 committed:
this replacing path_holder logic is moved to kolibri ZipContentView

66eli77 committed:
add nextQuestion and exercise basic styling

66eli77 committed:
Merge branch 'fork-richard' of https://github.com/66eli77/kolibri-exercise-perseus-plugin into fork-richard

66eli77 committed:
switch to prop-based instead of event-based communication between parent and children components.

66eli77 committed:
use userid to seededshuffle the questions

66eli77 committed:
move attemptProgress to perseus.vue

66eli77 committed:
add nextContent method that the behavior haven't been implemented yet

66eli77 committed:
modify setItemData for reusability.

66eli77 committed:
implemented how hint works. Maybe we should save the hint info in the database for a more robust precedure.

66eli77 committed:
pass masteryCriterion as prop

66eli77 committed:
add firstAttempt for immediate attempt-progress rendering

66eli77 committed:
dynamic attemptProgress based on passRatioM

66eli77 committed:
use the hinted field for attemptProgress

66eli77 committed:
style perseus

66eli77 committed:
add attemptprogress componenet

66eli77 committed:
add svgs for hint, correct, incorrect

66eli77 committed:
more styling for perseus

66eli77 committed:
modify the hint style to align with the final mockup

66eli77 committed:
refined hint styling

66eli77 committed:
add progress message to attemptprogress and style next content btn

66eli77 committed:
style next content btn

indirectlylit commented on exercise_perseus_render/assets/src/vue/perseus.vue:
```diff
@@ -165,13 +191,47 @@
+            this.complete = check.correct;
+            this.correct = this.hinted || !this.firstAttempt ? false : check.correct;
+            this.$parent.$emit('checkanswer', this.correct, this.complete, this.firstAttempt, this.hinted);
+            console.log('PPPPPPP:::: ', this.passNum);
```
mind clearing out all the debugging code before we merge this in?

66eli77 replied:
haha

indirectlylit commented on exercise_perseus_render/assets/src/vue/index.vue:
```diff
@@ -1,36 +1,102 @@
 <script>
 
-  const defaultQuestion = require('./question.json');
+  const ss = require('seededshuffle');
```
is this in package.json?

66eli77 replied:
the question.json is a dummy json Richard created for testing, now we just use the real one from Jordan.
seededshuffle is a seeded randomize library. I use the user_id as the seed.

indirectlylit replied:
I meant to say - has the `seededshuffle` library been added to the dependencies list (defined in this plugin's `package.json` file)?

66eli77 replied:
good point, I should add it here instead of kolibri

indirectlylit commented on exercise_perseus_render/assets/src/vue/perseus.vue:
```diff
@@ -1,33 +1,35 @@
+      <div id="hintsarea"></div>
       <div style="clear: both;"></div>
     </div>
+    <div id="answer-area-wrap">
```
whitespace cleanup below

indirectlylit commented on exercise_perseus_render/assets/src/vue/perseus.vue:
```diff
@@ -1,33 +1,35 @@
         <button v-else disabled id="scratchpad-not-available">{{ $tr("notAvailable") }}</button>
     </div>
     <div v-el:perseus-container id="perseus-container">
-    </div>
```
looks like an unclosed `div` here?

66eli77 replied:
oops, the indentation isn't aligned. The div is closed.

indirectlylit commented on exercise_perseus_render/assets/src/vue/perseus.vue:
```diff
@@ -165,13 +191,47 @@
+      attemptProgress() {
+        if (this.pastattempts) {
+          if (this.pastattempts.length > this.passRatioM){
+            const lastFiveAttempts = this.pastattempts.slice(0, this.passRatioM);
```
this variable should be renamed

indirectlylit commented on exercise_perseus_render/assets/src/vue/perseus.vue:
```diff
@@ -137,21 +142,42 @@
       // Has an answer been submitted?
       empty: true,
+      // how far from passing the exercise?
+      passNum: 4,
```
I think it would be much cleaner to implement `passNum` as a computed property rather than updating it in `attemptProgress`.

66eli77 replied:
You cannot pass computed as props in Vue

indirectlylit replied:
of course you can - let's chat if there's confusion

indirectlylit commented on exercise_perseus_render/assets/src/vue/perseus.vue:
```diff
@@ -165,13 +191,47 @@
+          if (!check.empty) {
+            this.complete = check.correct;
+            this.correct = this.hinted || !this.firstAttempt ? false : check.correct;
+            this.$parent.$emit('checkanswer', this.correct, this.complete, this.firstAttempt, this.hinted);
```
Could you add a comment explaining what this line does?

Calling `this.$parent` is very unusual, let alone `this.$parent.$emit`.

66eli77 replied:
why Calling this.$parent is very unusual?

indirectlylit replied:
It's uncommon for a child component to directly reference the parent component. It goes against the composition rule of only communicating with events: https://vuejs.org/guide/components.html#Composing-Components

indirectlylit replied:
see related comment here: https://github.com/learningequality/kolibri/pull/565#discussion_r84396546

indirectlylit reviewed (changes requested):
Hey - mind making a few quick updates to this?

indirectlylit commented on exercise_perseus_render/assets/src/vue/perseus.vue:
```diff
@@ -1,33 +1,35 @@
+                </button>
+                <button v-else id="hint-btn" disabled>
+                  <svg class="lightbulb disabled" src="./lightbulb_black.svg"></svg>{{ $tr("noMoreHint") }}
+                </button>
```
we should use the core `icon-button` component for all buttons

66eli77 committed:
add portrait mode for attempt-progress

66eli77 committed:
add top level wrapper to perseus
fix indentation alignment
change lastFiveAttempts to lastMAttempts
clean console log

66eli77 committed:
add seededshuffle here instead of in kolibri

66eli77 committed:
use icon-button

66eli77 committed:
delete trailing comma in package.json

66eli77 committed:
better formatting.

66eli77 committed:
abandon attemptProgress, use computed porterty instead

indirectlylit commented on exercise_perseus_render/assets/src/vue/perseus.vue:
```diff
@@ -196,7 +192,7 @@
             this.complete = check.correct;
             this.correct = this.hinted || !this.firstAttempt ? false : check.correct;
-            this.$parent.$emit('checkanswer', this.correct, this.complete, this.firstAttempt, this.hinted);
+            this.$emit('checkanswer', this.correct, this.complete, this.firstAttempt, this.hinted);
```
may have removed `$parent` by accident while we were talking?

66eli77 committed:
add $parent back to emit checkanswer

66eli77 committed:
use $nav-portrait-height

66eli77 committed:
check pastattempts for fresh new exercise attempt

66eli77 committed:
check user kind.
for now, exercise is only fully functional for LEANER.

66eli77 committed:
since we limit the pastattempts to 10 most recent attampts, we need to use totalattempts to loop the questions.

indirectlylit reviewed (approved):
all right!
